include/ghex/structured/cubed_sphere/field_descriptor.hpp

```diff
@@ -0,0 +1,504 @@
+                        }
+
+                        /** @brief accesses field at specified local coordinate
+                          * @param coord 4-dimensional array (component, x, y, z) in local coordinate system
+                          * @return reference to the value in the field */
+                        GT_FUNCTION
+                        T& data(const coordinate_type& coord) const noexcept {
+                            // make data memory coordinates from local coordinates
+                            const coordinate_type data_coord = coord + m_data_is.m_offset;
+                            // dot product with strides to compute address
+                            const auto memory_location =
+                                m_data_is.m_strides[0]*data_coord[0] +
+                                m_data_is.m_strides[1]*data_coord[1] +
+                                m_data_is.m_strides[2]*data_coord[2] +
+                                m_data_is.m_strides[3]*data_coord[3];
+                            return *reinterpret_cast<T*>(
+                                reinterpret_cast<char*>(m_data_is.m_ptr) + memory_location);
+                        }
+                    };
+
+                private: // members
+                    domain_id_type  m_dom_id;        /// domain id
+                    int             m_c;             /// cube size
+                    bool            m_is_vector;     /// is this data describing a vector field 
+                    value_type*     m_data;          /// pointer to data
+                    coordinate_type m_dom_first;     /// global coordinate of first domain cell
+                    coordinate_type m_offsets;       /// offset from beginning of memory to the first domain cell
+                    coordinate_type m_extents;       /// extent of memory (including halos)
+                    device_id_type  m_device_id;     /// device id
+                    strides_type    m_byte_strides;  /// memory strides in bytes
+
+                public: // ctors
+                    /** @brief construct field_descriptor from existing data
+                      * @tparam Array 3-dimensional array-like type
+                      * @param dom domain descriptor
+                      * @param data pointer to data
+                      * @param offsets (x,y,z)-offset of first phyisical value (not halo) from the orign of the memory
+                      * @param extents (x,y,z)-extent of the memory (including buffer regions)
+                      * @param n_components number of components in this field
+                      * @param is_vector whether this field is a vector or a collection of scalars
+                      * @param d_id device id */
+                    template<typename Array>
+                    field_descriptor(const domain_descriptor& dom, value_type* data,
+                                     const Array& offsets, const Array& extents, unsigned int n_components,
```

Comment:
This field type uses components, too. This field type, though, should be identical to the structured grid fields, right? Should we just keep one field descriptor for both cases? Or are the packing/unpacking functions specific to inflated-cube? Could we reuse at least part of the code?

Reply:
I gave it a try - it changed a lot of the code - but I believe it became more maintainable